### DISCUSSION TITLE
How to update title in editor via js?

### DISCUSSION THREAD

**Discussion by oeyoews** [TiddlyWiki/TiddlyWiki5#8293] [Q&A]
I want to implement the function of ai automatically generating a title based on the text content, but I encountered a problem.

```js
// $tw.wiki.findDraft('New Tiddler')
$tw.wiki.setText("Draft of 'New Tiddler'", "draft.title", null, "new ai title test")
```

But if I manually update the title in edit mode, this line of code will not work.

## Steps

* Open https://tiddlywiki.com
* create a tiddler
* Enter the above line of code in the console. You can see that the title `New Tiddler` has changed to `new ai title test`
* Manually modify the title of this tiddler
* Run the line of code again in the console, and the title name has not changed.

@Jermolene Do you have any ideas?

**Comment by pmario** [comment]
- The problem is, that the title is not set, if the title-input field has focus. 
- If you move the focus to the text field and then run your code, it does update

**Comment by pmario** [comment]
It seems to be a "solved" problem since v5.1.23, where the [EditWidget](https://tiddlywiki.com/#EditWidget) got 4 new parameters. One of them is `refreshTitle`, which refreshes the title of a tiddler, even if it has focus. 

As a workaround you need to edit the [$:/core/ui/EditTemplate/title](https://tiddlywiki.com/#%24%3A%2Fcore%2Fui%2FEditTemplate%2Ftitle) and set that parameter to eg: 

```
<$edit-text field="draft.title"
  class="tc-titlebar tc-edit-texteditor"
  focus={{{ [{$:/config/AutoFocus}match[title]then[true]] ~[[false]] }}}
  tabindex={{$:/config/EditTabIndex}}
  cancelPopups="yes"
  refreshTitle="$:/state/editor/refreshTitle"
/>
```

**See the last parameter.** 

Then do this:

```
$tw.wiki.setText("Draft of 'New Tiddler'", "draft.title", null, "new ai title test")
$tw.wiki.setText("$:/state/editor/refreshTitle", "text", null, "")
```

@Jermolene ... IMO we should add a "state" tiddler to the core edit-template, so plugin-authors are able to activate the "title refresh" function if needed.

**Reply by pmario** [reply]
Some discussion about the purpose of the `refreshTitle` parameter, with the initial [implementation](https://github.com/Jermolene/TiddlyWiki5/pull/4458#issuecomment-623133131)

>> Saq Imtiaz asked
>>
>>is the purpose of the refreshTitle parameter to edit-text, to be able to refresh the displayed input value without triggering a refreshSelf and loosing focus?
>
>BTC responded
>
>the first purpose is that it makes it possible to update the input field with what is stored in the tiddler that's being edited. we're setting that tiddler's content through `action-setfield` widgets and the input field doesn't update if we don't tell him to do so afterwards
